python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -146,16 +159,198 @@ def get_last_updated_timestamp(record: Optional[EventLogRecord]) -> Optional[flo
+            "Expected partitions_def to be time-windowed.",
+        )
+
+        last_update_time_lower_bound = get_last_update_time_lower_bound(
+            freshness_cron=freshness_cron,
+            freshness_cron_timezone=freshness_cron_timezone,
+            current_timestamp=current_timestamp,
+            lower_bound_delta=lower_bound_delta,
+        )
+        expected_partition_key = get_latest_complete_partition_key(
+            freshness_cron=freshness_cron,
+            current_timestamp=current_timestamp,
+            freshness_cron_timezone=freshness_cron_timezone,
+            partitions_def=partitions_def,
+        )
+        latest_record = retrieve_latest_record(
+            instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+        )
+        update_timestamp = get_last_updated_timestamp(latest_record)
+        passed = (
+            update_timestamp is not None
+            and update_timestamp >= last_update_time_lower_bound.timestamp()
+        )
+
+        return AssetCheckResult(
+            passed=passed,
+            description=get_description_for_freshness_check(
+                passed,
+                update_timestamp,
+                last_update_time_lower_bound,
+                current_timestamp,
+                expected_partition_key,
+            ),
+            severity=severity,
+        )
+
+    return the_check
+
+
+def get_description_for_freshness_check(
+    passed: bool,
+    update_timestamp: Optional[float],
+    last_update_time_lower_bound: datetime.datetime,
+    current_timestamp: float,
+    expected_partition_key: Optional[str],
+) -> str:
+    check.invariant(
+        (passed and update_timestamp is not None) or not passed,
+        "Should not be possible for check to pass without a record.",
+    )
+    update_time_delta_str = (
+        pendulum.duration(seconds=current_timestamp - update_timestamp).in_words()
+        if update_timestamp
+        else None
+    )
+    last_update_time_lower_bound_delta_str = pendulum.duration(
+        seconds=current_timestamp - last_update_time_lower_bound.timestamp()
+    ).in_words()
+    return (
+        f"Partition {expected_partition_key} is fresh. Expected a record for the partition within the last {last_update_time_lower_bound_delta_str}, and found one last updated {update_time_delta_str} ago."
+        if passed and expected_partition_key
+        else f"Partition {expected_partition_key} is overdue. Expected a record for the partition within the last {last_update_time_lower_bound_delta_str}."
+        if not passed and expected_partition_key
+        else f"Asset is fresh. Expected a record within the last {last_update_time_lower_bound_delta_str}, and found one last updated {update_time_delta_str} ago."
+        if passed and update_timestamp
+        else f"Asset is overdue. Expected a record within the last {last_update_time_lower_bound_delta_str}."
+    )
```

Comment:
I think we should be talking about "updates" instead of "records" here. There can be an observation record that doesn't indicate the asset was updated.

That said, for time-partitioned assets, I think the wording "Partition X is overdue. Expected an update" is a little misleading, because in most cases they're expecting it to be the first arrival of that partition, rather than an update to an existing partition.  The prior code uses "Partition X arrived on time" vs. "Partition X is Y minutes late", which I think avoids these difficulties.

Reply:
> I think we should be talking about "updates" instead of "records" here. There can be an observation record that doesn't indicate the asset was updated.

Make total sense 👍🏻

At some point (outside scope of this diff) we should clarify "observation" "materialization" "record" "update" ontology and document internally and externally

Reply:
https://linear.app/dagster-labs/issue/FOU-128/refresh-and-document-ontology-around-materialization-observation